internal/scheduler/scheduler.go

```diff
@@ -177,30 +198,38 @@ func (s *Scheduler) start(ctx context.Context) {
 			event.WriteSysEvent(ctx, op, "scheduling loop shutting down", "server id", s.serverId)
 			return
 		case <-timer.C:
-			repo, err := s.jobRepoFn()
-			if err != nil {
-				event.WriteError(ctx, op, err, event.WithInfoMsg("error creating job repo"))
-				break
-			}
+		case <-s.runNow:
+		}
+		start := time.Now()
+		repo, err := s.jobRepoFn()
+		if err != nil {
+			event.WriteError(ctx, op, err, event.WithInfoMsg("error creating job repo"))
+			break
+		}
 
-			runs, err := repo.RunJobs(ctx, s.serverId, job.WithRunJobsLimit(s.runJobsLimit))
-			if err != nil {
-				event.WriteError(ctx, op, err, event.WithInfoMsg("error getting jobs to run from repo"))
-				break
-			}
+		runs, err := repo.RunJobs(ctx, s.serverId, job.WithRunJobsLimit(s.runJobsLimit))
+		if err != nil {
+			event.WriteError(ctx, op, err, event.WithInfoMsg("error getting jobs to run from repo"))
+			break
+		}
 
-			for _, r := range runs {
-				err := s.runJob(ctx, &wg, r)
-				if err != nil {
-					event.WriteError(ctx, op, err, event.WithInfoMsg("error starting job"))
-					if _, inner := repo.FailRun(ctx, r.PrivateId, 0, 0); inner != nil {
-						event.WriteError(ctx, op, inner, event.WithInfoMsg("error updating failed job run"))
-					}
+		for _, r := range runs {
+			err := s.runJob(ctx, &wg, r)
+			if err != nil {
+				event.WriteError(ctx, op, err, event.WithInfoMsg("error starting job"))
+				if _, inner := repo.FailRun(ctx, r.PrivateId, 0, 0); inner != nil {
+					event.WriteError(ctx, op, inner, event.WithInfoMsg("error updating failed job run"))
 				}
 			}
 		}
 
-		timer.Reset(s.runJobsInterval)
+		switch {
+		case s.runNowRequested.After(start):
```

Comment:
I'm not sure I follow, what I was trying to do here was not have a number of threads blocking on the `RunNow` call each writing to the `runNow` chan.  So I had the `RunNow` method attempt to write, but if nothing is receiving on that chan, the default is called and it exits. Setting the timestamp is just me making sure we honor the `RunNow` call and ensure that after calling it, the scheduler will run as soon as possible, the timer should only be set to zero if something called `RunNow` while the scheduling loop was not in the select and actually doing the scheduling. 
Happy to have a quick call to discuss if I am missing something.
As a side not the job will not run more than it is scheduled to, all this is doing is triggering the scheduling loop to look for jobs to run.

Reply:
Oh you are right.  I forgot that channels are default size 0 not 1 so if there is no receiver on the other end of the channel it wont send in this case.  Instead of tracking the runNowRequested timestamps, what do you think about making the runNow channel size 1?  I get a little uneasy when we start doing timestamp comparisons across goroutines to try to handle race conditions.

Reply:
Good idea, thanks